---
title: Identification of latest bug in predictCoding() function
output:
  html_document:
    df_print: paged
knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_dir = "html_output"
    )
  })
---

## Loading needed packages

```{r message = FALSE}
library(VariantAnnotation)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(BSgenome.Hsapiens.UCSC.hg38)
library(DT)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
txdb <- keepStandardChromosomes(txdb)
```


## Loading the VCF files, subsetting to ATP13A2

```{r message=FALSE}
## Location of the SNP in TAF4
snp <- GRanges("chr20", IRanges(62003269))

vcfObjectsFile <- file.path("data", "unique_SNPs_vs_all.Rds")
vcfs <- readRDS(vcfObjectsFile)
vcf <- vcfs[["ATP13A2"]]
vcf <- keepStandardChromosomes(vcf)
vcf <- vcf[rowRanges(vcf) == snp]
rowRanges(vcf)
```



## Locating the SNPs associated with TAF4
```{r message=FALSE}
cds <- locateVariants(vcf, txdb, CodingVariants())

data.frame(cds$QUERYID, cds$TXID, cds$GENEID)
```

## Finding it's influence on the coding sequence
```{r message=FALSE}
aa <- predictCoding(vcf, txdb, Hsapiens)

aa
data.frame(aa$QUERYID, aa$TXID, aa$GENEID)
aa
```


```{r}
library(org.Hs.eg.db)

keytypes(org.Hs.eg.db)
columns(org.Hs.eg.db)

genes <- c("TAF4","GBA1","PCDHB2","SPESP1")
genes

ezid <- select(org.Hs.eg.db,genes,"ENTREZID","SYMBOL")$ENTREZID


## All of the genes are in the txdb database
all(ezid %in% keys(txdb,'GENEID'))


tx <- transcriptsBy(txdb, "gene")

tx[ezid]

ezid %in% names(tx)

gene.id <- unique(select(txdb, aa$TXID[is.na(aa$GENEID)], "GENEID", "TXID")$GENEID)

ex <- exonsBy(txdb, "gene")

ex[txid]

ex <- exons(txdb)

ex

keytypes(txdb)

tt <- ex[ex$exon_id %in% select(txdb, gene.id, "EXONID", "GENEID")$EXONID]

tt
seqnames(tt)
start(tt)
end(tt)
end(tt) - start(tt)
strand(tt)
tx

tx[unique(select(txdb, aa$TXID[is.na(aa$GENEID)], "GENEID", "TXID")$GENEID)]

tx

cds$GENEID

txdb
```
